design/datamover.md

```diff
@@ -0,0 +1,48 @@
+- The interface for Data Mover should support moving data associated with resources **other than** Persistent Volumes
+  - Persistent Volumes are an immediate need, and will likely drive most of the original design discussions.
+  - Ex. a Data Mover for Kubevirt would move VM data to backup storage. The interface should be generic enough to support this use case.
+- The interface for Data Mover should enable publishing it’s specific configuration parameters for users to have fine-grained control over the movement workflow
+  - Examples:
+    - Data mover for vsphere volume can configure block vs file mount usage
+    - Kubevirt mover can configure export file type
+    - Image mover can specify image format type
+- End-user interaction with data mover should be exposed through the existing backup/restore operations
+
+## Non-Goals
+- This implementation will **NOT** change the existing Velero backup/restore workflow.
+  - It is possible in the future that this interface can be used to replace some legacy Velero implementations, but that is **NOT** a goal for this design/implementation.
+- This design of the interface should not include any specific details about (aside from samples):
+  - A mover type
+    - Although we are focused on volumes for phase 1, the interface should have zero knowledge of the underlying data that needs to be moved.
+  - Mover implementations
+    - Ex. How a mover handles volumes at the block vs filesystem level is not relevant to the overarching interface design
+
+## Use Cases
+- Data Mover for Persistent Volume Snapshots
+  - For a given snapshot type, a capable data mover should be able to clone the snapshot data into some backup target.
+  - The most urgent use case for this today is CSI snapshots of volumes where the given CSI driver does not guarantee durability and portability. A user creates a CSI snapshot of a volume and wishes to clone this snapshot data into some backup target. The user then would like to be able restore a volume from this snapshot that exists in the backup target.
+- User wants to consume third party vendor data movers
+  - Example: Dell Power Protect has their own data mover to handle PVs and would like to integrate this into the native Velero backup workflow as opposed to it being a separate and disconnected process.To do this, Velero’s data mover interface has to support the ability for a vendor to bring their own data mover and have the Velero backup/restore workflow trigger it’s process.
+- User wants to move data associated with resources other than Persistent Volumes
+  - Example: User takes a backup of a kubevirt VM. A Kubevirt data mover would be capable of exporting the VM and pushing it to backup storage.
+  - Data mover interface should be capable of defining mover strategies for any type of kubernetes resource.
+
+## Workflow
```

Comment:
I am a bit confused by the workflow and how it relates to existing backup and restore.  In order to be compatible with the existing velero flows wouldn't we need something like the following:

Backup/offload
1. Velero is configured to use CSI Snapshots.
2. User requests a backup that includes some PVs
3. Backup is performed (this results in the creation of some VolumeSnapshot objects)
4. DataMover offloads the VolumeSnapshot object by: cloning the VolumeSnapshot to a temporary volume, copying the data from the temporary volume to a BSL, and deleting the temporary volume.

Load/Restore
1. VolumeSnapshot is recreated in the cluster by: creating a temporary volume in the cluster, loading data from the BSL into the volume, creating a VolumeSnapshot from the volume, and deleting the temporary volume.
2. Restore is performed using the recreated VolumeSnapshot as a source for the data.

Reply:
For restore, why recreate the VolumeSnapshot? The new restore process should be able to restore the data to PV/PVC which is exactly the same as when it is backup'd.

Reply:
For CSI case, PV restoration is not reusing the same PV, because PV is already backed up by snapshot, so the PV will be recreated from snapshot, but the PV name may still be same.